crates/uv-workspace/src/pyproject_mut.rs

```diff
@@ -503,13 +503,38 @@ pub fn add_dependency(
     deps: &mut Array,
     has_source: bool,
 ) -> Result<ArrayEdit, Error> {
-    // Find matching dependencies.
     let mut to_replace = find_dependencies(&req.name, Some(&req.marker), deps);
-    match to_replace.as_slice() {
+
+    let edit = match to_replace.as_slice() {
         [] => {
+            // determine the dependency list is sorted prior to
+            // adding the new dependency; the new dependency list
+            // will be sorted only when the original list is sorted
+            // so that users' custom dependency ordering is preserved.
+            let sorted = deps
+                .clone()
+                .into_iter()
+                .collect::<Vec<_>>()
+                .windows(2)
+                .all(|w| w[0].to_string() <= w[1].to_string());
+
             deps.push(req.to_string());
             reformat_array_multiline(deps);
-            Ok(ArrayEdit::Add(deps.len() - 1))
+
+            if sorted {
```

Comment:
For the sorted case: Since we don't parse change already requirements, we shouldn't deserialize and then `to_string()` them. Instead, we should convert the new requirement to a string and use this as comparison, finding the index with `partition_point` .

Reply:
let me look into this, it's slightly tricky since `.to_string()` on each of the dependencies yields strings of the form `\n    \"CacheControl[filecache]>=0.14,<0.15\"`; there may be a way to strip those relatively easily.

Reply:
There is a method in the `toml` crate that allows you to get the contained string value, instead of the string representation of the parsed field with decoration that `.to_string()` seems to use.

Reply:
trimming the empty space + `\n` turned out to be pretty straight-forward but the formatting part is turning out to be trickier...
```
        Value(
            String(
                Formatted {
                    value: "sentry-sdk",
                    repr: "\"sentry-sdk\"",
                    decor: Decor {
                        prefix: "\n    ",
                        suffix: empty,
                    },
                },
            ),
        ),
        Value(
            String(
                Formatted {
                    value: "pydantic",
                    repr: "default",
                    decor: Decor {
                        prefix: " ",
                        suffix: empty,
                    },
                },
            ),
        ),
```
where `reformat_array_multiline` seems to be operating under the assumption that only the last value of `deps` has an empty prefix.
```
            deps.get_mut(deps.len() - 1)
                .unwrap()
                .decor_mut()
                .set_prefix("\n    ");
```
it's possible to go with the partition solution still if we set the prefix manually for the inserted value and the last value, but it may be cleaner to just keep the current implementation?

Reply:
In this case we should update `reformat_array_multiline`. I'm confident that we should sort just by `d.as_str()` (without parsing and serializing), since this is the value that the user sees in the file.

Reply:
sounds good, i'll take a look again sometime.